test/test_wand_data.cpp

```diff
@@ -0,0 +1,109 @@
+#define CATCH_CONFIG_MAIN
+#include "catch2/catch.hpp"
+
+#include <functional>
+
+#include "test_common.hpp"
+
+#include "pisa_config.hpp"
+#include "index_types.hpp"
+#include "query/queries.hpp"
+#include "wand_data_range.hpp"
+#include "scorer/score_function.hpp"
+
+using namespace pisa;
+
+TEST_CASE("wand_data_range") {
+    typedef wand_data_range<64, 1024, bm25> WandTypeRange;
+    typedef wand_data<bm25, WandTypeRange> WandType;
+    using Scorer = bm25;
+
+    binary_freq_collection collection(PISA_SOURCE_DIR "/test/test_data/test_collection");
+    binary_collection      document_sizes(PISA_SOURCE_DIR "/test/test_data/test_collection.sizes");
+    WandType          wdata_range(document_sizes.begin()->begin(),
+                              collection.num_docs(),
+                              collection);
+    SECTION( "Precomputed block-max scores" ) {
+        size_t i = 0;
+        for (auto const &seq: collection) {
+            if(seq.docs.size() >= 1024) {
+                auto max = wdata_range.max_term_weight(i);
+                auto w   = wdata_range.getenum(i);
+                for (auto j = 0; j < seq.docs.size(); ++j) {
```

Comment:
Nope. `j` means nothing to me. Remember we talked about what collection is and how to iterate over it? If you obfuscate variables like this, it's difficult for anyone to understand what you're doing, especially someone new to the code.

Another thing is that you're not actually using `j` anyway. You do it only to move the iterators. Which is not a great way to do it. Why not define iterators and increment them? Or better yet, use ranges:

```cpp
for (auto&& [docid, freq] : ranges::view::zip(seq.docs, seq.freqs)) {...}
```